### PR TITLE
Pv2 1179 pv2 1162 persist datalock data and re run data locks when apprenticeship changes

### PR DESCRIPTION
**Pull request by samuel-dev-dba**

### CODE REVIEW

File: src/SFA.DAS.Payments.DataLocks.Application.UnitTets/Services/DataLockEventProcessorTest.cs [hunk 1]
```diff
@@ -96,50 +107,57 @@ public async Task TestCreatesDataLockChangedEventWhenNewFailure()
...
+        public async Task TestCreatesMultipleEventsForPeriodsAndTypes(DataLockEvent failureEvent)
         {
             // arrange
 
 
 
             // change of dlock code
-            var oldTT2P5 = new DataLockFailureEntity {Id = 1, DeliveryPeriod = 5, TransactionType = TransactionType.Completion, EarningPeriod = new EarningPeriod { Period = 5, DataLockFailures = new List<DataLockFailure> {new DataLockFailure {DataLockError = DataLockErrorCode.DLOCK_06}}}};
-            var newTT2p5 = new EarningPeriod {Period = 5, Amount = 1, DataLockFailures = new List<DataLockFailure> {new DataLockFailure {DataLockError = DataLockErrorCode.DLOCK_04}}};
+            var oldTT2P5 = new DataLockFailureEntity { Id = 1, DeliveryPeriod = 5, TransactionType = TransactionType.Completion, EarningPeriod = new EarningPeriod { Period = 5, DataLockFailures = new List<DataLockFailure> { new DataLockFailure { DataLockError = DataLockErrorCode.DLOCK_06 } } } };
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.DataLocks.Application.UnitTets/Services/DataLockEventProcessorTest.cs]: the variable names are a little confusing (I think there is a commented description about each one but it's not the easiest to follow)

File: src/SFA.DAS.Payments.DataLocks.Application.UnitTets/Services/DataLockEventProcessorTest.cs [hunk 2]
```diff
@@ -96,50 +107,57 @@ public async Task TestCreatesDataLockChangedEventWhenNewFailure()
         }
 
         [Test]
-        public async Task TestCreatesMultipleEventsForPeriodsAndTypes()
+        [TestCaseSource(nameof(GetFailureEvents))]
+        public async Task TestCreatesMultipleEventsForPeriodsAndTypes(DataLockEvent failureEvent)
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.DataLocks.Application.UnitTets/Services/DataLockEventProcessorTest.cs]: maybe for later, but this test method is far too big. Not sure we are actually gaining anything by testing the three statuses together.

File: src/SFA.DAS.Payments.DataLocks.Application/Repositories/ApprenticeshipRepository.cs [hunk 3]
```diff
@@ -134,7 +139,24 @@ public async Task UpdateCurrentlyPausedApprenticeship(ApprenticeshipPauseModel a
...
+        {
+            var apprenticeshipEarning = await dataContext.EarningEvent
+                .Include(x => x.Periods)
+                .Include(x => x.PriceEpisodes)
+                .Where(x => x.Ukprn == ukprn && 
+                            x.LearnerUln == uln && 
+                            x.ContractType == ContractType.Act1 &&
+                            x.EventType == eventType)
+                .OrderByDescending(x => x.Id)
+                .Take(1)
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.DataLocks.Application/Repositories/ApprenticeshipRepository.cs]: do you need to take 1 and then get first or default? Is that just to limit what gets selected in SQL and then further making sure in code we get one?

Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Application/Repositories/ApprenticeshipRepository.cs]: You can't do use first or default when you use OrderBy so forced to use Take 1

File: src/SFA.DAS.Payments.DataLocks.Application.UnitTets/Services/DataLockEventProcessorTest.cs (hunk 2 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Application.UnitTets/Services/DataLockEventProcessorTest.cs]: I will leave this, for now, this is part of Carl changes

File: src/SFA.DAS.Payments.DataLocks.Application.UnitTets/Services/DataLockEventProcessorTest.cs (hunk 1 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Application.UnitTets/Services/DataLockEventProcessorTest.cs]: The test was written by you so I will leave the variable names, for now, can change it later.

**Pull request merged**